src/System.Private.CoreLib/src/System/Environment.CoreCLR.cs

```diff
@@ -0,0 +1,248 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Win32;
+
+namespace System
+{
+    public static partial class Environment
+    {
+        public static int CurrentManagedThreadId => Thread.CurrentThread.ManagedThreadId;
+
+        // Terminates this process with the given exit code.
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        private static extern void _Exit(int exitCode);
+
+        public static void Exit(int exitCode) => _Exit(exitCode);
+
+        public static extern int ExitCode
+        {
+            [MethodImpl(MethodImplOptions.InternalCall)]
+            get;
+            [MethodImpl(MethodImplOptions.InternalCall)]
+            set;
+        }
+
+        // Note: The CLR's Watson bucketization code looks at the caller of the FCALL method
+        // to assign blame for crashes.  Don't mess with this, such as by making it call 
+        // another managed helper method, unless you consult with some CLR Watson experts.
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern void FailFast(string message);
+
+        // This overload of FailFast will allow you to specify the exception object
+        // whose bucket details *could* be used when undergoing the failfast process.
+        // To be specific:
+        //
+        // 1) When invoked from within a managed EH clause (fault/finally/catch),
+        //    if the exception object is preallocated, the runtime will try to find its buckets
+        //    and use them. If the exception object is not preallocated, it will use the bucket
+        //    details contained in the object (if any).
+        //
+        // 2) When invoked from outside the managed EH clauses (fault/finally/catch),
+        //    if the exception object is preallocated, the runtime will use the callsite's
+        //    IP for bucketing. If the exception object is not preallocated, it will use the bucket
+        //    details contained in the object (if any).
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern void FailFast(string message, Exception exception);
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        public static extern void FailFast(string message, Exception exception, string errorMessage);
+
+        private static string GetEnvironmentVariableCore(string variable)
```

Comment:
`GetEnvironmentVariableCore`, `GetEnvironmentCharArray` and `SetEnvironmentVariableCore` can be in shared under  `<ItemGroup Condition="$(TargetsWindows) or '$(FeaturePal)'=='true'">`. (There is an item group like that already.)